misc-scripts/xref_mapping/XrefParser/UniProtParser/Extractor.pm

```diff
@@ -0,0 +1,645 @@
+  # you want to force it to only ever look for one simply drop the /g
+  # modifier from the regex match.
+
+  Readonly my $taxon_db_entry
+    => qr{
+           # Database qualifier. Chances are the list of
+           # allowed characters will change should DBs
+           # other than NCBI ever become supported here.
+           ( [A-Za-z_]+ )
+
+           \s*  # just in case
+           =
+           \s*  # same
+
+           # Taxon ID. This is almost certainly NCBI-specific.
+           ( [0-9]+ )
+       }msx;
+  Readonly my $evidence_code_list
+    => qr{
+           # As of October 2018, this syntax is not declared in
+           # UniProt-KB User Manual yet frequently encountered in data
+           # files.  Use named sequences for curly brackets to avoid
+           # excessive escaping as well as for better readability.
+           \N{LEFT CURLY BRACKET}
+           \s*
+           [^\N{RIGHT CURLY BRACKET}]+
+           \s*
+           \N{RIGHT CURLY BRACKET}
+           \s*
+       }msx;
+  my @ox_captures
+    = ( $ox_line =~ m{
+                       $taxon_db_entry
+                       \s*
+
+                       # Optional things (e.g. evidence codes) we do
+                       # not need now but must parse on the off chance
+                       # someone decides e.g. that curly braces
+                       # constitute quotes so it is okay to have a
+                       # semicolon between them
+                       (?:
+                         $evidence_code_list
+                         | [^;]+
+                       )?
+
+                       # End of record
+                       ;
+                   }gmsx );
+
+  my @extracted_taxon_codes;
+
+  # Reminder: if you change the number of capture groups above
+  # remember to adapt both the variable list *and* the third argument
+  # to splice().
+ TAXON_ENTRY:
+  while ( my ( $db_qualifier, $taxon_code ) = splice( @ox_captures, 0, 2 ) ) {
+
+    if ( ( ! defined $db_qualifier )
+         || ( ! exists $supported_taxon_database_qualifiers{$db_qualifier} ) ) {
+      # Abort on malformed or new database qualifiers
+      croak "Cannot use taxon-DB qualifier '${db_qualifier}'";
+    }
+    elsif ( ! $supported_taxon_database_qualifiers{$db_qualifier} ) {
+      # Known but of no interest. Ignore it.
+      next TAXON_ENTRY;
+    }
+
+    if ( ! defined $taxon_code ) {
+      croak "Failed to extract taxon code from:\n\t${ox_line}";
+    }
+
+    # FIXME: further processing?
+
+    push @extracted_taxon_codes, {
+                                  'db_qualifier' => $db_qualifier,
+                                  'taxon_code'     => $taxon_code,
+                                };
+
+  }
+
+  return \@extracted_taxon_codes;
+}
+
+
+sub _record_has_all_needed_fields {
+  my ( $self ) = @_;
+
+  # Only check mandatory fields
+  my @needed_fields = grep { $prefixes_of_interest{$_} } %prefixes_of_interest;
```

Comment:
I think you could use a `keys` here or It'll do interesting things with every other element.

Reply:
D'oh. Harmless in this context but you are of course right, grep simply takes a list.